data_diff/database.py

```diff
@@ -110,6 +112,38 @@ def md5_to_int(self, s: str) -> str:
         "Provide SQL for computing md5 and returning an int"
         ...
 
+    @abstractmethod
+    def _query(self, sql_code: str) -> list:
+        "Send query to database and return result"
+        ...
+
+
+class ThreadedDatabase(Database):
+    """Access the database through singleton threads.
+
+    Used for database connectors that do not support sharing their connection between different threads.
+    """
+
+    def __init__(self, thread_count=1):

```

Comment:
In a follow-up you'll allow specifying the concurrency in the CLI? 👍🏻

Reply:
Even with this default though, it does seem to use concurrency?

```
[10:39:11] INFO - . Diffing segment 1/12 of size 90908 and 90908
[10:39:11] INFO - . Diffing segment 8/12 of size 90909 and 90909
[10:39:11] INFO - . Diffing segment 9/12 of size 90909 and 90909
[10:39:11] INFO - . Diffing segment 10/12 of size 90909 and 90909
[10:39:11] INFO - . Diffing segment 11/12 of size 90909 and 90909
[10:39:11] INFO - . Diffing segment 12/12 of size 2 and 2
[10:39:11] INFO - . Diffing segment 7/12 of size 90909 and 90909
[10:39:11] INFO - . Diffing segment 5/12 of size 90909 and 90909
[10:39:11] INFO - . Diffing segment 6/12 of size 90909 and 90908
[10:39:11] INFO - . Diffing segment 4/12 of size 90909 and 90909
[10:39:11] INFO - . Diffing segment 2/12 of size 90909 and 90909
[10:39:11] INFO - . Diffing segment 3/12 of size 90909 and 90909
```

Why isn't it basically single-threaded with `thread_count=1`?

Reply:
Yes, the CLI will pass an argument for that, once we figure out the interface.

The database access is single-threaded (when =1), but the algorithm itself is still threaded. That's necessary for taking advantage of thread_count>1. I could limit the algorithm to 1 thread too in this case, but I'm not sure that it matters.

Reply:
I think it matters because in that case I'd expect to see the output in order if threads=1